inc/export/class-endpoint.php

```diff
@@ -0,0 +1,292 @@
+		wp_raise_memory_limit( 'analytics' );
+		set_time_limit( 0 );
+
+		// Manually stream the response, as there is likely too much to hold in memory at once.
+		header( 'X-Accel-Buffering: no' );
+
+		// Check accept header for format.
+		$parsed_header = Utils\parse_accept_header( $request->get_header( 'accept' ) );
+		$accept_type = Utils\find_best_accept_header_match( $parsed_header, [
+			'application/json',
+			'text/csv',
+		] );
+
+		$format = $accept_type === 'text/csv' ? 'csv' : $request->get_param( 'format' );
+
+		// Set content type header.
+		if ( $format === 'json' ) {
+			header( 'Content-Type: application/json; charset=' . get_option( 'blog_charset' ), true );
+		} else {
+			header( 'Content-Type: text/csv; charset=' . get_option( 'blog_charset' ), true );
+		}
+
+		// Default part of the query to get all events for the current site.
+		$main_query = [
+			'bool' => [
+				'filter' => [
+					[
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						'range' => [
+							'event_timestamp' => [
+								'gte' => $dates['start'],
+								'lt' => $dates['end'],
+							],
+						],
+					],
+				],
+			],
+		];
+
+		$total_query = [
+			'query' => $main_query,
+			'size' => 0,
+		];
+
+		// Ensure we're tracking the total hits for ES 7.3+ so our slices aren't too large.
+		if ( version_compare( Utils\get_elasticsearch_version(), '7.2', '>' ) ) {
+			$total_query['track_total_hits'] = true;
+		}
+
+		// The first query returns just the total number of items
+		// so the results can be sliced later.
+		$total_results = Utils\query( $total_query );
+		$total = $total_results['hits']['total']['value'] ?? $total_results['hits']['total']; // ES 7 compat.
+
+		// Set total found results header.
+		header( sprintf( 'X-WP-Total: %d', $total ) );
+
+		// Determine the chunk size.
+		$per_page = $request->get_param( 'per_page' ); // Should not be more than 10000.
+		$per_page = max( 1, $per_page );
+
+		// Elasticsearch slices need at least 2 pages and no more than 1024 by default.
+		$total_pages = (int) ceil( $total / $per_page );
+		$total_pages = min( 1024, max( $total_pages, 2 ) );
+
+		// Collect all the fields we'll need ahead of time for CSV output.
+		$fields = [];
+
+		// Begin output.
+		if ( $format === 'json' ) {
+			echo "[\n";
+		} else {
+			// Fetch all available columns across all indices if we're sending a csv, we need these for the 1st line.
+			$index_mappings = Utils\query( [], [
+				'ignore_unavailable' => 'false',
+				'include_defaults' => 'false',
+				'filter_path' => '-*.mappings.*keyword,-*.mappings._*,-*.mappings.*.*.*',
+			], '_mapping/field/*', 'GET' );
+
+			foreach ( $index_mappings as $mapping ) {
+				$fields = array_merge( $fields, array_keys( $mapping['mappings'] ) );
+				$fields = array_unique( $fields );
+			}
+
+			sort( $fields );
+
+			// Write columns headings line.
+			fputcsv( fopen( 'php://output', 'w' ), $fields ) . "\n";
```

Comment:
You're reopening output handles each time here; you should open the handle once. Also, after each call here, you should also call `fflush()`

Reply:
Okey doke, might need some updates where this pattern is used elsewhere 😬